ecpy_hqc_legacy/instruments/drivers/driver_tools.py

```diff
@@ -147,6 +146,80 @@ def wrapper(self, *args, **kwargs):
     return decorator
 
 
+class InstrJob(object):
+    """Object returned by instrument starting a long running job.
+
+    Parameters
+    ----------
+    condition_callable : Callable
+        Callable taking no argument and indicating if the job is complete.
+
+    expected_waiting_time : float
+        Expected waiting time for the task to complete in seconds.
+
+    cancel : Callable, optional
+        Function to cancel the task.
+
+    """
+    def __init__(self, condition_callable, expected_waiting_time, cancel):
+        self.condition_callable = condition_callable
+        self.expected_waiting_time = expected_waiting_time
+        self.cancel = cancel
+        self._start_time = time.time()
+
+    def wait_for_completion(self, break_condition_callable=None, timeout=15,
+                            refresh_time=1):
+        """Wait for the task to complete.
+
+        Parameters
+        ----------
+        break_condition_callable : Callable, optional
+            Callable indicating that we should stop waiting.
+
+        timeout : float, optional
+            Time to wait in seconds in addition to the expected condition time
+            before breaking.
+
+        refresh_time : float, optional
+            Time interval at which to check the break condition.
+
+        Returns
+        -------
+        result : bool
+            Boolean indicating if the wait succeeded of was interrupted.
+
+        """
+        while True:
+            remaining_time = (self.expected_waiting_time -
+                              (time.time() - self._start_time))
+            if remaining_time < 0:
+                break
+            time.sleep(min(refresh_time, remaining_time))
+            if break_condition_callable():
+                return False
```

Comment:
Don't you want to call the cancel here ? otherwise you have to at least call it in the source logic (it is never called right now if I'm correct)

Reply:
Not here because I want user to be able to decide whet to do if there is a timeout or an interruption.